### ISSUE TITLE
Release Ubuntu Xenial Server 16.04 LTS Packages

### ISSUE THREAD

**Issue by felixbuenemann** [phusion/passenger#1807]
The passenger apt repositories are currently missing packages for the current Ubuntu Server LTS version.

**Comment by OnixGH** [comment]
As noted on the [forum](https://groups.google.com/forum/#!topic/phusion-passenger/rdatUaQxWX8): The build for 16.04 is still on our TODO list. The current aim is to get it out in June at the latest.

In the meantime, Passenger can be installed on 16.04 as a gem or from source.

**Comment by felixbuenemann** [comment]
Guess I'll have to use passenger standalone and proxy to it from nginx in the meantime.

**Comment by felixbuenemann** [comment]
Btw. the passenger packages shipped with Xenial are weird. They only packages the dependencies for apache, so the passenger package is missing the `passenger` binary.

**Comment by OnixGH** [comment]
@felixbuenemann I'm not sure I understand your remark about standalone, if you have Passenger installed as a gem or from source you still have the same options as when it was installed through apt (e.g. `passenger-install-nginx`)?

**Comment by felixbuenemann** [comment]
I am aware of that, but I though I could just run passenger-standlone, given that xenial ships a fairly recent version of passenger (5.0.27). I'm no fan of installing my own binaries that do not get ciritical updates through the package manager, as they tend to get forgotten which is especially bad for public services like webservers…

**Comment by OnixGH** [comment]
Ah, so what you're saying is that the temporary setup of a (standalone / proxying) Nginx is to have it ride along on updates through the package manager, instead of relying on gem updates.

Would it make any difference to you if Passenger had some kind of mechanism to log or warn about critical security updates (i.e. helping to ensure it doesn't get overlooked if for instance, it was a gem install that doesn't benefit from apt)?

**Comment by felixbuenemann** [comment]
Yes, that's what I'm saying.

I'm not sure if a different warning mechanism would be helpful, given that all our tooling for automating updates across all out servers is based on apt.

I tried building nginx using `passenger-install-nginx-module`, but it's lacking lot's of modules and functionality present in the ubuntu packages so it doesn't work with out default nginx and vhosts configs.

Is the source tarball for the nginx packages available for download? Then I could compile nginx with the same modules as the ubuntu packages.

**Comment by OnixGH** [comment]
Yeah, a warning system wouldn't help in terms of automation, just as a backup / reminder to make sure critical updates aren't forgotten.

The `passenger-install-nginx-module` actually just downloads Nginx straight from their website, so we don't have/need a special source tarball. You can download any Nginx version and any module you need and run `passenger-install-nginx-module` with option 2 (customize), point it to the downloaded Nginx source and add any --(module-options).

That gets added to the default configure command (which is something like `./configure --prefix='/<path.to>/nginx' --with-http_ssl_module --with-http_gzip_static_module --with-http_stub_status_module --with-cc-opt=-Wno-error --with-ld-opt='' --add-module='/<path.to>/passenger/src/nginx_module'`)

**Comment by felixbuenemann** [comment]
Yes, but I still have to hunt down all the right modules and versions.

Anyways, I found the included ones at https://github.com/phusion/passenger_apt_automation/tree/master/debian_specs/nginx/modules so I can download that and look at `nginx -V` on a trusty system to figure out what I need.

**Comment by ernetas** [comment]
Any estimates on when packages for Ubuntu 16.04 will be available for Passenger Enterprise?

**Comment by OnixGH** [comment]
@ernetas same estimate as given in this thread.

**Comment by ernetas** [comment]
Any chance you would then add passenger-enterprise specifications to https://github.com/phusion/passenger_apt_automation scripts (correct me if I'm wrong, but that repo seems to include only Debian package specs for Passenger Enterprise, but not the rake tasks, etc.)?

**Comment by felixbuenemann** [comment]
The apt automation doesn't yet have a base image for Xenial, so you won't get far.

**Comment by ernetas** [comment]
Not true. I have forked the phusion/baseimage and rebuilt it from `ubuntu:xenial` a long time ago, so building the open source packages using the new image didn't take 2 hours to figure the rest... I hosted the open source packages at: 
https://trys.eu/apt/passenger (note: this is an apt repo, so you don't have to download each file individually, just add `deb https://trys.eu/apt/passenger/ /` to `/etc/apt/sources.list` and run `apt-key adv --keyserver keyserver.ubuntu.com --recv 3611EABD0F03ECC87598A8C08E76E753C4E95086 && apt-get update` (if you're reading this less than a day after this was posting, you may need to retry the apt-key command, as I just pushed the public key to keyserver.ubuntu.com, so it takes some time to propogate to all keyserver.ubuntu.com servers)).

Sadly, I need Enterprise packages, not OSS, so this is still of no use to me...

If that would make things any faster for you, I could make some PRs introducing some changes needed to have support for Xenial (mostly it's just some updates/changes (e.g. fixing `/sbin/setuser`, installing `dh-systemd`) to the `passenger_apt_automation_buildbox` image and adding Ruby 2.3 support to debian_specs).

**Comment by felixbuenemann** [comment]
Last time I tried building for Xenial, pbuilder was missing the base image tgz, so it aborted. After that I didn't look into it any further.

Thanks for posting the packages, I'll give them a try.

**Comment by ernetas** [comment]
I have pushed some of the changes that I had to do to build Passenger for Xenial at https://github.com/ernetas/passenger_apt_automation/tree/xenial, if that helps anyhow. Note that I used my own built image (replace `registry.example.com...` with yours), that had some modifications that are not in that repo (I'm too lazy to recall all of them).

**Comment by FooBarWidget** [comment]
It isn't just about making a 16.04 build box. Our packages have to be compatible with upstream Debian packages, so we need to study how the Debian upstream packaging changed between 15.10 and 16.04 and merge those changes with our own while maintaining compatibility with older distributions.

**Comment by udf2457** [comment]
> we need to study how the Ubuntu packaging

My understanding is that 15.x was a migration release and that 16.x is merely enforcing what was previously optional in 15.x (e.g. systemd was default but optional in 15.x, but now in 16.x it is mandatory with the fallback option removed).

> while maintaining compatibility with older distributions.

Surely the whole point of being able to specify a distribution to track in the APT config is you don't need to worry about maintaining compatability with a legacy distribution if you publish your packages as 16.04 specific ?

**Comment by felixbuenemann** [comment]
@udf2457 I think @FooBarWidget is not only talking about the systemd vs. upstart change, but also about which nginx modules are built into the various system nginx packages (basic, full, extras etc.).

**Comment by FooBarWidget** [comment]
> Surely the whole point of being able to specify a distribution to track in the APT config is you don't need to worry about maintaining compatability with a legacy distribution if you publish your packages as 16.04 specific ?

But we don't just publish 16.04 packages. We publish packages for Ubuntu 12.04, 14.04, and now 16.04. We also publish packages for Debian 7 and 8. We generate all these packages from a single source in order to avoid duplicate effort, but this also means that we have to be extremely careful about merging changes from Ubuntu so that we don't break compatibility with older distribution versions.

Anyway, 16.04 support is on its way.

**Comment by carukc** [comment]
It's great to hear that it's on the way.  Thank you f or the update.

**Comment by ehannes** [comment]
> The current aim is to get it out in June at the latest.

Now it's June :) Any release date?

**Comment by ehannes** [comment]
> In the meantime, Passenger can be installed on 16.04 as a gem or from source.

If I choose to install Passenger as a gem, how would I switch back to the apt version when that's available? (Is it worth the trouble?)

**Comment by FooBarWidget** [comment]
> Now it's June :) Any release date?

It's being worked on right now. We ask for your patience. Packaging is not a simple matter, it takes a lot of time to get right.

**Comment by wurde** [comment]
Just learned of the APT repository provided by passenger, then promptly disappointing it wasn't supporting the current Ubuntu LTS. Will this be an ongoing problem with each new release or is the solution being worked on going to enable timely releases?

**Comment by FooBarWidget** [comment]
@wurde We have been working on Ubuntu 16.04 packages for some time now. It is scheduled for release this week.

**Comment by ernetas** [comment]
It sounds like Phusion would really benefit from adding extra component to the apt repository like "experimental". Similar to Docker's repository. I mean, noone expects stable packages to be released very fast and it's nice to have a very critical attitude towards releasing new package versions, but it's also very painful for some of us to be blocked by something like application server for starting to even test newer OSes, especially if OSes you need to support does not depend on you, but on your clients.

**Comment by FooBarWidget** [comment]
While we do not have an "experimental" repository, our packaging work is [done in the public](https://github.com/phusion/passenger_apt_automation). If you wish to have prerelease/in-development packages then you can take our in-progress packaging work and build packages yourself.

**Comment by wurde** [comment]
I'd add that a lot of our codebase includes *architecture as code* and has provisions to automatically roll forward to the most recent releases of software including the underlying Ubuntu OS LTS. So in deciding what technology choices, Passenger has a real chance to be the best solution with it's APT and nginx integration. As it is, installation from source works fine despite having to purge any existing nginx/passenger packages.

**Comment by felixbuenemann** [comment]
Upgraded from custom install in /opt/nginx to the official Xenial packages today.

Worked without a hitch. Thanks, @FooBarWidget and @OnixGH!

One thing I noticed is that the `nginx.service` systemd unit uses QUIT signal, which is fine for the usual setup, but causes trouble with unix sockets, because it will leave stale sockets around which cause nginx restart to fail. So if you are binding to unix sockets, make sure to edit the systemd unit to use TERM.

Details: http://unix.stackexchange.com/questions/164866/nginx-leaves-old-socket

**Comment by OnixGH** [comment]
@felixbuenemann thanks for letting us know, we'll have to investigate that.

**Comment by OnixGH** [comment]
@felixbuenemann reproduced the issue and switched the script from QUIT to TERM in our package builders, so 5.0.30 and beyond be OK: 
https://github.com/phusion/passenger_apt_automation/commit/c3d2940e8390c57ea953db4e55b8990b89367e2e | https://github.com/phusion/passenger_rpm_automation/commit/f868d8ea60a72a6a7f65d2361c7d4eb3f2cb5ddf

Thanks!
